File: bindings/heeczer-js/scripts/parity.mjs
import { readFile, readdir } from "node:fs/promises";
import path from "node:path";
import { fileURLToPath } from "node:url";

import { HeeczerClient } from "../dist/index.js";

const scriptDir = path.dirname(fileURLToPath(import.meta.url));
const repoRoot = path.resolve(scriptDir, "../../..");
const fixtureDir = path.resolve(
  process.env.HEECZER_PARITY_FIXTURE_DIR ??
    path.join(repoRoot, "core/schema/fixtures/events/valid"),
);
const referenceDir = process.env.HEECZER_PARITY_REFERENCE_DIR;
const baseUrl = process.env.HEECZER_PARITY_BASE_URL;

if (!referenceDir || !baseUrl) {
  throw new Error(
    "HEECZER_PARITY_REFERENCE_DIR and HEECZER_PARITY_BASE_URL are required",
  );
}

const client = new HeeczerClient({
  baseUrl,
  retry: { attempts: 3, backoffMs: 50 },
});

const fixtureNames = (await readdir(fixtureDir))
  .filter((name) => name.endsWith(".json"))
  .sort((left, right) => left.localeCompare(right));

if (fixtureNames.length === 0) {
  throw new Error(`no valid fixtures found in ${fixtureDir}`);
}

const failures = [];
for (const fixtureName of fixtureNames) {
  const eventPath = path.join(fixtureDir, fixtureName);
  const referencePath = path.join(
    referenceDir,
    `${path.basename(fixtureName, ".json")}.json`,
  );
  const event = JSON.parse(await readFile(eventPath, "utf8"));
  const expected = (await readFile(referencePath, "utf8")).trimEnd();
  const response = await client.testScorePipeline({ event });
  const actual = JSON.stringify(response.score);
  if (actual !== expected) {
    failures.push(`${fixtureName}: score JSON differed from Rust reference`);
  }
}

if (failures.length > 0) {
  throw new Error(failures.join("\n"));
}

console.log(`JS SDK parity passed for ${fixtureNames.length} fixture(s)`);

File: bindings/heeczer-py/scripts/parity.py
from __future__ import annotations

import asyncio
import json
import os
from pathlib import Path

from heeczer import HeeczerClient


def _required_env(name: str) -> str:
    value = os.environ.get(name)
    if not value:
        raise RuntimeError(f"{name} is required")
    return value


async def main() -> None:
    repo_root = Path(__file__).resolve().parents[3]
    fixture_dir = Path(
        os.environ.get(
            "HEECZER_PARITY_FIXTURE_DIR",
            repo_root / "core" / "schema" / "fixtures" / "events" / "valid",
        )
    )
    reference_dir = Path(_required_env("HEECZER_PARITY_REFERENCE_DIR"))
    base_url = _required_env("HEECZER_PARITY_BASE_URL")

    fixtures = sorted(fixture_dir.glob("*.json"))
    if not fixtures:
        raise RuntimeError(f"no valid fixtures found in {fixture_dir}")

    failures: list[str] = []
    async with HeeczerClient(
        base_url=base_url,
        retry={"attempts": 3, "backoff_ms": 50},
    ) as client:
        for fixture_path in fixtures:
            event = json.loads(fixture_path.read_text(encoding="utf-8"))
            expected = (reference_dir / f"{fixture_path.stem}.json").read_text(
                encoding="utf-8"
            ).rstrip()
            response = await client.test_score_pipeline(event=event)
            actual = json.dumps(response["score"], separators=(",", ":"))
            if actual != expected:
                failures.append(
                    f"{fixture_path.name}: score JSON differed from Rust reference"
                )

    if failures:
        raise AssertionError("\n".join(failures))
    print(f"Python SDK parity passed for {len(fixtures)} fixture(s)")


if __name__ == "__main__":
    asyncio.run(main())
